src/service/layers-impl.js

```diff
@@ -0,0 +1,876 @@
+    for (let l = this; l !== stopAt; l = l.getParentLayer()) {
+      const position = l.getOffsetFromParent();
+      // Calculate the scrolled position. If the element has offset 200, and
+      // the element is scrolled 150, then the scrolled position is just 150.
+      // Note that the scrolled position shouldn't take into account the scroll
+      // position of this LayoutElement, so we've already compensated in
+      // declaring x and y.
+      x += position.left - l.getScrollLeft();
+      y += position.top - l.getScrollTop();
+    }
+
+    return positionLt(x, y);
+  }
+
+  /**
+   * Returns the absolute offset position of the element relative to the layer
+   * represented by opt_parent (or opt_parent's parent layer, if it is not a
+   * layer itself). This remains constant, regardless of the scrolled position
+   * of any layer in between.
+   *
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getOffsetPosition(opt_parent) {
+    let x = 0;
+    let y = 0;
+
+    // Find the layer to stop measuring at. This is so that you can find the
+    // relative position of an element from some parent element, say the
+    // position of a slide inside a carousel, without any further measurements.
+    let stopAt = opt_parent ? LayoutElement.getParentLayer(opt_parent) : null;
+
+    for (let l = this; l !== stopAt; l = l.getParentLayer()) {
+      const position = l.getOffsetFromParent();
+      // Add up every offset position in the ancestry.
+      x += position.left;
+      y += position.top;
+    }
+
+    return positionLt(x, y);
+  }
+
+  /**
+   * Dirties the element, so the next size/position calculation will remeasure
+   * the element.
+   */
+  requestRemeasure() {
+    this.needsRemeasure_ = true;
+  }
+
+  /**
+   * Dirties the layer, so the next position calculation will remeasure the
+   * scroll positions.
+   */
+  requestScrollRemeasure() {
+    this.needsScrollRemeasure_ = true;
+  }
+
+  /**
+   * Remasures the element's size and offset position. This traverse as high as
+   * possible in the layer tree to remeasure as many elements as possible in
+   * one go. This is necessary both from a performance standpoint, and to
+   * ensure that any calculation uses the correct value, since layer in the
+   * ancestry may have been dirtied.
+   *
+   * No matter what, though, the current element will be remeasured.
+   */
+  remeasure() {
+    let layer = this;
+
+    // Find the root-est layer that's dirty, and remeasure from there.
+    for (let p = this.getParentLayer(); p; p = p.getParentLayer()) {
+      if (p.needsRemeasure_) {
+        layer = p;
+      }
+    }
+
+    layer.remeasure_();
+  }
+
+  /**
+   * Remeasures the element, and all children, since this element was marked dirty.
+   *
+   * @param {!PositionDef=} opt_relativeTo A performance optimization used when
+   *     recursively measuring the child nodes of the layer.
+   */
+  remeasure_(opt_relativeTo) {
```

Comment:
Semantics, but we shouldn't need a concept of "remeasure" anymore. Just dirty flags and "get position/size/scroll" requests.

Reply:
Externally, yes (and I think the mutate element APIs will just dirty the flags).

Internally, we have to remeasure to make the flags clean again.